### DISCUSSION TITLE
node_modules in server side?

### DISCUSSION THREAD

**Discussion by addlistener** [vikejs/vike#571] [Discussion]
I am using https://github.com/brillout/vite-plugin-ssr/tree/main/examples/react-router/

And I get a lot of `Cannot use import statement outside a module` when importing `node_modules/*/*.js` files to server. I guess it's because node.js runtime run the raw file of `import xxx` instead of the transpiled `require(xxx)`

After hacking `type="module"` in the package.json of those node_modules it works. But it's annoying, since the deps are deeply nested and I constantly get these errors.

My question
1. is there a unified way of changing all packages to `type="module"`?
2. Am I using vite-plugin-ssr the right way? I guess a lot of 3rd party libs are using  `import xxx` and what should I do? Should I add some transpilation to server files?

**Comment by brillout** [comment]
> And I get a lot of `Cannot use import statement outside a module` when importing `node_modules/*/*.js` files to server.

What's the full stack trace?

**Comment by antun** [comment]
@brillout, I also ran into this same error when I tried to use this the **examples/redux** template with a specific 3rd-party package. (In my case the 3rd-party package is [react-live-runner](https://www.npmjs.com/package/react-live-runner?activeTab=explore), which seems to have all the modern JS exports set-up, but still throws this error.) I wrote a [question on StackOverflow](https://stackoverflow.com/questions/76442590/vite-ssr-getting-error-cannot-use-import-statement-outside-a-module-when-rende) before I found this discussion, and that has a bunch of details.

In my case, the full error is:

```
4:55:46 PM [vite] Error when evaluating SSR module /pages/index.page.jsx: failed to import "react-live-runner"
|- /Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-live-runner/dist/index.modern.js:1
import{useRunner as e}from"react-runner";export*from"react-runner";import t,{useState as r,useEffect as n,Fragment as a,useCallback as l,useRef as o,useMemo as c,createContext as s,useContext as i}from"react";import p from"react-simple-code-editor";import m,{Prism as u}from"prism-react-renderer";function g(){return g=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e},g.apply(this,arguments)}function y(e,t){if(null==e)return{};var r,n,a={},l=Object.keys(e);for(n=0;n<l.length;n++)t.indexOf(r=l[n])>=0||(a[r]=e[r]);return a}const d=["initialCode","transformCode"],h=t=>{let{initialCode:a="",transformCode:l}=t,o=y(t,d);const[c,s]=r(a),{element:i,error:p}=e(g({code:l?l(c):c},o));return n(()=>{s(a)},[a]),{element:i,error:p,code:c,onChange:s}},f={plain:{color:"#ffffff",backgroundColor:"#282c34"},styles:[{types:["comment","block-comment","prolog","doctype","cdata"],style:{color:"#b2b2b2"}},{types:["property","number","function-name","constant","symbol","deleted"],style:{color:"#5a9bcf"}},{types:["boolean"],style:{color:"#ff8b50"}},{types:["tag"],style:{color:"#fc929e"}},{types:["string","attr-value"],style:{color:"#8dc891"}},{types:["punctuation"],style:{color:"#88c6Be"}},{types:["selector","char","builtin","inserted"],style:{color:"#d8dee9"}},{types:["function"],style:{color:"#79b6f2"}},{types:["operator","entity","url","variable"],style:{color:"#d7deea"}},{types:["keyword"],style:{color:"#c5a5c5"}},{types:["atrule","class-name"],style:{color:"#fac863"}},{types:["important"],style:{fontWeight:"400"}},{types:["bold"],style:{fontWeight:"700"}},{types:["italic"],style:{fontStyle:"italic"}},{types:["entity"],style:{cursor:"help"}},{types:["namespace"],style:{opacity:.7}}]},b=["children","language","theme","Prism","padding","noWrapper","noWrap","className","style"],C=e=>{let{children:r,language:n="jsx",theme:l=f,Prism:o=u,padding:c=10,noWrapper:s,noWrap:i,className:p,style:d}=e,h=y(e,b);/*#__PURE__*/return t.createElement(m,{code:r||"",language:n,Prism:o,theme:l},({className:e,style:r,tokens:n,getLineProps:l,getTokenProps:o})=>{const m=n.map((e,r)=>/*#__PURE__*/t.createElement(a,{key:r},/*#__PURE__*/t.createElement("span",l({line:e}),e.map((e,r)=>/*#__PURE__*/t.createElement("span",o({token:e,key:r})))),"\n"));return s?m:/*#__PURE__*/t.createElement("pre",g({className:p?`${e} ${p}`:e,style:g({},r,{margin:0,padding:c,whiteSpace:i?"pre":"pre-wrap"},d)},h),m)})},v=["defaultValue","value","language","theme","Prism","highlight","padding","onChange"],E=e=>{let{defaultValue:n,value:a,language:s="jsx",theme:i=f,Prism:m,highlight:u,padding:d=10,onChange:h}=e,b=y(e,v);const[E,P]=r(n||""),k=void 0!==a,W=l(e=>/*#__PURE__*/t.createElement(C,{language:s,theme:i,Prism:m,noWrapper:!0},e),[s,i,m]),j=o(h);j.current=h;const x=l(e=>{k||P(e),null==j.current||j.current(e)},[k]),O=c(()=>g({},i.plain,b.style),[i.plain,b.style]);/*#__PURE__*/return t.createElement(p,g({},b,{highlight:u||W,padding:d,value:k?a:E,onValueChange:x,style:O}))},P=s({}),k=()=>i(P),W=["children","code","language","theme"],j=e=>{let{children:r,code:n="",language:a="jsx",theme:l=f}=e,o=y(e,W);const{element:c,error:s,code:i,onChange:p}=h(g({initialCode:n},o));/*#__PURE__*/return t.createElement(P.Provider,{value:{element:c,error:s,code:i,onChange:p,language:a,theme:l}},r)},x=e=>{const{code:r,language:n,theme:a,onChange:l}=k();/*#__PURE__*/return t.createElement(E,g({value:r,language:n,theme:a,onChange:l},e))},O=["Component"],w=e=>{let{Component:r="div"}=e,n=y(e,O);const{element:a}=k();/*#__PURE__*/return t.createElement(r,n,a)},N=e=>{const{error:r}=k();return r?/*#__PURE__*/t.createElement("pre",e,r):/*#__PURE__*/t.createElement(t.Fragment,null)};export{C as CodeBlock,E as CodeEditor,P as LiveContext,x as LiveEditor,N as LiveError,w as LivePreview,j as LiveProvider,f as defaultTheme,k as useLiveContext,h as useLiveRunner};
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at Object.compileFunction (node:vm:360:18)
    at wrapSafe (node:internal/modules/cjs/loader:1084:15)
    at Module._compile (node:internal/modules/cjs/loader:1119:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
    at Module.load (node:internal/modules/cjs/loader:1033:32)
    at Function.Module._load (node:internal/modules/cjs/loader:868:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:169:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Promise.all (index 0)
```

I split the **examples/redux** code into a repo and added the package that's causing problems here:

https://github.com/antun/vite-ssr-redux-react-live-runner-example

**Reply by brillout** [reply]
Have you seen https://vite-plugin-ssr.com/broken-npm-package? This will most likely solve your problem. Let me know whether it does.

**Reply by antun** [reply]
@brillout thank you - that does solve the first problem I ran into above. I'd actually come across the `ssr.noExternal` property while doing some further debugging, but didn't think to use it for this issue. Basically, if update the `vite.config.js` file to:

```
import react from '@vitejs/plugin-react'
import ssr from 'vite-plugin-ssr/plugin'

export default {
  plugins: [react(), ssr()],
  ssr: {
    noExternal: ['react-live-runner', 'react-runner']
  }
}
```

... then I'm able to both develop (by running `vite`) and build (by running `vite build`).

The issue that I'm now facing is that if I enable SSG (by changing `ssr()` to `ssr({ prerender: true })` in `vite.config.js`, then developing works, but `vite build` fails with a stack trace when it gets to the `pre-rendering HTML...` stage, and the error doesn't give me a whole lot to go on:

```
vite-plugin-ssr v0.4.131 pre-rendering HTML...
Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.
    at Ib (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:72:155)
    at W (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:73:89)
    at Ib (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:67:482)
    at W (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:73:89)
    at Ib (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:67:482)
    at W (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:73:89)
    at Jb (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:76:98)
    at Kb (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:75:140)
    at W (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:73:345)
    at Ib (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:71:304)
```

**Reply by brillout** [reply]
Can you publish a reproduction?

**Reply by antun** [reply]
@brillout yes, I just pushed up some changes to my test repo here:

https://github.com/antun/vite-ssr-redux-react-live-runner-example

Specifically:

* I enabled prerender
* I added the noExternal settings to avoid the problems with the react-live-runner and react-runner packages
* I reduced the case to just including the `<LiveEditor>` component - it seems that adding that causes `vite build` to break with the `Error: Element type is invalid` error. It looks like there's something about that component it doesn't like.

The steps to reproduce are `npm install && vite build`.

**Reply by brillout** [reply]
Insert the following:

```js
// node_modules/react/cjs/react.production.min.js

// ...

// Insert this at this end of the file
exports.createElement = debugEl(exports.createElement)
function debugEl(fn) {
  return (...args) => {
    const el = args[0]
    if (typeof el === 'object' && !el.$$typeof && !el.getModuleId) {
      console.error('Error: INVALID REACT ELEMENT:');
      console.error(el)
      console.error(new Error('INVALID REACT ELEMENT, ORIGIN:'));
      globalThis.process?.exit?.(1)
    }
    return fn(...args)
  }
}
```

```js
// node_modules/react/cjs/react-jsx-runtime.production.min.js

// ...

// Insert this at this end of the file
exports.jsx = debugEl(exports.jsx)
function debugEl(fn) {
  return (...args) => {
    const el = args[0]
    if (typeof el === 'object' && !el.$$typeof && !el.getModuleId) {
      console.error('Error: INVALID REACT ELEMENT:');
      console.error(el)
      console.error(new Error('INVALID REACT ELEMENT, ORIGIN:'));
      globalThis.process?.exit?.(1)
    }
    return fn(...args)
  }
}
```

It should help you get to the bottom of it.

(While I could reproduce using Gitpod, I couldn't reproduce locally on my machine. I'll try again and further dig later.)

**Reply by brillout** [reply]
> It should help you get to the bottom of it.

With that debugging information, you'll be able to discover further (nested) broken npm pacakges. Add them to `ssr.noExternal` and that should fix the problem.

Let me whether that works. If it doesn't then I'll further dig into it.

**Reply by brillout** [reply]
(Reproduction with pnpm for further reference: https://github.com/brillout/vite-ssr-redux-react-live-runner-example.)

**Reply by antun** [reply]
What is it that `pnpm` does differently from `npm` - is that why it's working for you locally?

--

UPDATE: I tried switching to `pnpm` and it gives *much* better output for this type of debugging, e.g.

> Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import '/Users/antun/git/ui-pilot/ui-pilot-ui/node_modules/@mui/lab/LoadingButton' is not supported resolving ES modules imported from /Users/antun/git/ui-pilot/ui-pilot-ui/build/server/entries/src_pages_index_index-page.mjs

... and:

> Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import '/Users/antun/git/ui-pilot/ui-pilot-ui/node_modules/@mui/material/utils' is not supported resolving ES modules imported from /Users/antun/git/ui-pilot/ui-pilot-ui/build/server/entries/src_pages_index_index-page.mjs

my `ssr.noExternal` now contains a bunch of packages - I'm still debugging some issues as to why it's not generating, but this is how many I'm excluding in my actual project, not the simple test case:

```
  ssr: {
    noExternal: [
      '@reduxjs/toolkit',
      '@mui/material',
      '@mui/lab',
      '@mui/material/styles',
      'stylis',
      'react-split-pane',
      'react-runner',
      'react-simple-code-editor',
      'prism-react-renderer'
    ]
  },
```

Do people typically have to exclude these? Things like `@mui/material` are pretty widely-used, so I'm surprised they're not compliant.

--

Also in my [simple text case](https://github.com/antun/vite-ssr-redux-react-live-runner-example), I'm still somewhat stuck because while using `ssr.noExternal: ['react-live-runner', 'react-runner', 'react-simple-code-editor', 'prism-react-renderer']` allows me to run `vite build` successfully, `vite` _now_ fails with this error:

```
2:58:12 PM [vite] Error when evaluating SSR module /pages/index.page.jsx: failed to import "/node_modules/react-live-runner/dist/index.modern.js"
|- ReferenceError: exports is not defined
    at eval (/Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/react-simple-code-editor/lib/index.js:64:23)
    at instantiateModule (file:///Users/antun/git/scratch/vite-ssr-redux-react-live-runner-example/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:54405:15)
```

Previously I was able to run `vite`, but not `vite build`. I suppose I could try working-around this by having one `vite.config.js` for development and a different one for building. I still need to figure out what is failing in my full app though.

**Reply by antun** [reply]
Ah, that's very helpful - thanks!

Just posting some notes here in case someone else runs into this. After adding the debugging code to the React code, I'd get output like this:

```
Error: INVALID REACT ELEMENT:
{
  Prism: <ref *1> {
    util: {
      encode: [Function: encode],
...
```

Then I searched node_modules for references to that element, e.g. `grep -Rl '<Prism' node_modules`, and added those to the `noExternal` block of vite.config.ts. (Also I searched for just references to the package - e.g. `grep -Rl 'Prism' node_modules` - for some cases.

I'm still dealing with another different error now:

> SyntaxError: Named export 'ThemeProvider' not found. The requested module '@mui/material/styles/index.js' is a CommonJS module, which may not support all module.exports as named exports.

I do see that one has a couple of other workarounds listed on https://vite-plugin-ssr.com/broken-npm-package, and I'll try those. (The `ssr.noExternal` doesn't seem to work there.)

**Reply by brillout** [reply]
> Do people typically have to exclude these?

Yes, that isn't unusual. A lot of packages ship invalid code. That's very unfortunate and to fix the situation changes are needed that are outside the realm of VPS. The relevant entities are relunctant/slow to change, so it'll take some time until we iron this out.

> Then I searched node_modules for references to that element, e.g. `grep -Rl '<Prism' node_modules`, and added those to the `noExternal` block of vite.config.ts. (Also I searched for just references to the package - e.g. `grep -Rl 'Prism' node_modules` - for some cases.

I'd recommend to do the following instead.

A typical log will look like this:

```
Error: INVALID REACT ELEMENT:
{
  default: [Function: SomeComopnent]
}
Error: INVALID REACT ELEMENT, ORIGIN:
    at Object.createElement (node_modules/react/cjs/react.production.min.js:36:21)
    at E (dist/server/entries/pages_index-page.mjs:127:12)
    at Fb (node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:65:44)
    at Ib (node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:67:254)
    at W (node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:73:89)
    at Ib (node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:67:482)
    at W (node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:73:89)
    at Ib (node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:71:304)
    at W (node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:73:89)
    at Ib (node_modules/react-dom/cjs/react-dom-server-legacy.node.production.min.js:67:482)
```

> You can see the problem here: the *actual* React element is `[Function: SomeComopnent]` but a wrapper object is being passed instead. (The ESM default import object.)

Go to the position `dist/server/entries/pages_index-page.mjs:127:12` which will look like this:

```js
// Line 127, column 12:
/*...*/ createElement(someVar, /* ... */)
```

> You may see `jsx(/*...*/)` instead of `createElement(/*...*/)`.

You can temporarily fix the problem by replacing `someVar` with `someVar.default`.

To permanently fix the problem: go to where `someVar` is defined, which typically is an import of an npm package `some-npm-package`. Add `some-npm-package` to `ssr.noExternal` to permanently fix the problem.

(FYI this is the biggest problem currently with Vite/VPS (the only one actually...). I care a lot about fixing the situation but it's going to take some time. In the meantime, doing some manually digging like we're doing here should work, but if it doesn't then definitely let me know.)

**Reply by brillout** [reply]
Actually, if you set `ssr.noExternal: true` and use https://github.com/cyco130/vite-plugin-cjs-interop, does that solve all your problems? I believe it should... let me know!

There is still manual work to do though. Everytime you hit:

> SyntaxError: Named export 'someImport' not found. The requested module 'some-npm-pacakge/path/to/module.js' is a CommonJS module, which may not support all module.exports as named exports.

Then add to the package `some-npm-pacakge` to `vite-plugin-cjs-interop`:

```js
// vite.config.js
import { cjsInterop } from 'vite-plugin-cjs-interop'

export default {
  plugins: [
    cjsInterop({
      // List of dependencies that require interop
      dependencies: [
        'some-npm-package'
      ]
    })
  ]
}
```

(FYI, I think we can make `vite-plugin-cjs-interop` automatically detect such dependencies, so that it just works without manual work.)

Let me know how much `ssr.noExternal: true` degrades your dev performance. (Why it does is explained at https://vite-plugin-ssr.com/broken-npm-package#solution.)
